src/pykeen/models/unimodal/rgcn/decompositions.py

```diff
@@ -0,0 +1,398 @@
+# -*- coding: utf-8 -*-
+
+"""Various decompositions for R-GCN."""
+
+import logging
+from typing import Optional, Tuple, Union
+
+import torch
+from class_resolver import Resolver
+from torch import nn
+from torch.nn import functional
+
+__all__ = [
+    "Decomposition",
+    "decomposition_resolver",
+]
+
+
+def _reduce_relation_specific(
+    relation: int,
+    source: torch.LongTensor,
+    target: torch.LongTensor,
+    edge_type: torch.LongTensor,
+    edge_weights: Optional[torch.FloatTensor],
+) -> Union[Tuple[torch.LongTensor, torch.LongTensor, Optional[torch.FloatTensor]], Tuple[None, None, None]]:
+    """Reduce edge information to one relation.
+
+    :param relation:
+        The relation ID.
+    :param source: shape: (num_edges,)
+        The source node IDs.
+    :param target: shape: (num_edges,)
+        The target node IDs.
+    :param edge_type: shape: (num_edges,)
+        The edge types.
+    :param edge_weights: shape: (num_edges,)
+        The edge weights.
+
+    :return:
+        The source, target, weights for edges related to the desired relation type.
+    """
+    # mask, shape: (num_edges,)
+    edge_mask = edge_type == relation
+    if not edge_mask.any():
+        return None, None, None
+
+    source_r = source[edge_mask]
+    target_r = target[edge_mask]
+    if edge_weights is not None:
+        edge_weights = edge_weights[edge_mask]
+
+    # bi-directional message passing
+    source_r, target_r = torch.cat([source_r, target_r]), torch.cat([target_r, source_r])
+    if edge_weights is not None:
+        edge_weights = torch.cat([edge_weights, edge_weights])
+
+    return source_r, target_r, edge_weights
+
+
+class Decomposition(nn.Module):
```

Comment:
add abstract class and abstract methods?

Reply:
done in f6eb12735be98b7aeff7dd76b8d07a3d30fd6e8a